ClusterPool IPAM fixes & cleanups

tgraf opened the pull request.

sayboras commented on pkg/option/config.go:
```diff
@@ -2602,7 +2602,7 @@ func (c *DaemonConfig) Populate() {
 
 	switch c.IPAM {
-	case ipamOption.IPAMKubernetes:
+	case ipamOption.IPAMKubernetes, ipamOption.IPAMClusterPool:
```
:100: 

The failure in smoke test (quick install / experimential install) might be related to this commit. These smoke tests are using the docker images from docker hub, which didn't have this change. Both `conformance tests` with newly built docker images are passing, so I think it's fine.

sayboras commented on pkg/ipam/allocator/clusterpool/clusterpool.go:
```diff
@@ -61,35 +62,43 @@ type AllocatorOperator struct {
 		}
 		a.v4CIDRSet = v4Allocators
+	} else if len(operatorOption.Config.ClusterPoolIPv4CIDR) != 0 {
+		return errors.New("ClusterPool IPv4 CIDR can not be set if IPv4 is disabled")
```
nit: Can we use `operatorOption.ClusterPoolIPv4CIDR` instead ? And s/can/must/g ? And maybe fmt.Errorf() instead of errors.New() for consistency with above lines ?

sayboras commented on pkg/ipam/allocator/clusterpool/clusterpool.go:
```diff
@@ -61,35 +62,43 @@ type AllocatorOperator struct {
 		}
 		a.v6CIDRSet = v6Allocators
+	} else if len(operatorOption.Config.ClusterPoolIPv6CIDR) != 0 {
+		return errors.New("ClusterPool IPv6 CIDR can not be set if IPv6 is disabled")
```
nit: Can we use `operatorOption.ClusterPoolIPv6CIDR` instead ? And s/can/must/g ?

sayboras reviewed (approved):
LGTM. Some minor nits, feel free to address if you see fit :)

fristonio reviewed (approved):
LGTM :100:

christarazi commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -226,6 +226,8 @@ func (n *NodeDiscovery) UpdateCiliumNodeResource() {
 		return
 	}
 
+	log.WithField("node", nodeTypes.GetName()).Info("Creating or updating CiliumNode resource")
```
We should use [this field from the current list of established fields](https://github.com/cilium/cilium/blob/0db8cd04465e88144a8f3b93c057053c2b07faa6/pkg/logging/logfields/logfields.go#L32), because this is an `Info` msg which will always be in the logs. I think the general pattern I've noticed is we add/use fields from the list of established fields when it's an `Info` msg or higher. `Debug` msgs don't need an entry in the established fields list.

christarazi commented on operator/main.go:
```diff
@@ -288,6 +288,9 @@ func onOperatorStartLeading(ctx context.Context) {
 		nodeManager *allocator.NodeEventHandler
 	)
+
+	log.WithField("mode", option.Config.IPAM).Info("Initializing IPAM")
```
We should consider adding this to the list of established fields, since it's an `Info` msg.

christarazi commented on pkg/ipam/allocator/clusterpool/clusterpool.go:
```diff
@@ -89,7 +90,10 @@ func (a *AllocatorOperator) Init() error {
+	log.WithFields(logrus.Fields{
+		"ipv4CIDRs": operatorOption.Config.ClusterPoolIPv4CIDR,
+		"ipv6CIDRs": operatorOption.Config.ClusterPoolIPv6CIDR,
+	}).Info("Starting ClusterPool IP allocator")
```
Log fields here too.

christarazi reviewed (commented):
LGTM, just a few comments below, then it should be good to go for me.

tgraf committed:
ipam: Move pkg/ipam/allocator/operator to pkg/ipam/allocator/clusterpool

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
operator: Log when initializing the IPAM system

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Improve error messages when retrieving the CiliumNode resource

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Rename IPAMOperatorV[46]CIDR to ClusterPoolIPv[46]CIDR

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
nodediscovery: Log attempt to create or update CiliumNode resource

This help detect instances when creating or updating the resource is
delayed or hangs. It also helps validate the correct behavior of the
cluster-pool IPAM implementation.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Rename IPAMOperator to IPAMClusterPool

The operator is involved in various IPAM modes, a more specific name for
the ClusterPool mode makes the code more readable.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Derive PodCIDR requirements from enable-ipv[46] setting

There is no need to require special Helm logic to set
`k8s-require-ipv4-pod-cidr` correctly in combination with `ipam:
cluster-pool`. The existing `enable-ipv4` and `enable-ipv6` options are
more user friendly and ensure that cluster-pool also functions properly
without Helm.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Improve error message when PodCIDR is unavailable

When using ClusterPool IPAM, the reference to Kubernetes node is
misleading to users.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Rename GetNodeSpec() to WaitForNodeInformation()

The function has grown and supports both CiliumNode and Kubernetes Node
and thus the original name is misleading at this point.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
clusterpool: Error out if IPv6 is enabled but ClusterPool CIDR is not set

The existing logic reported an error if both IPv4 and IPv6 CIDRs were
not provided. This is insufficient when both IPv4 and IPv6 are enabled
but only an IPv4 ClusterPool CIDR is specified.

It lead to a situation where agents won't start up while waiting for the
CiliumNode to to provide an IPv6 address.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Log parameters when initializing ClusterPool

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
k8s: Set upper limit when backing off from node retrieval failure

Signed-off-by: Thomas Graf <thomas@cilium.io>
